src/node_native_module_env.cc

```diff
@@ -0,0 +1,203 @@
+            ToJsSet(context, can_be_required))
+      .FromJust();
+  info.GetReturnValue().Set(result);
+}
+
+void NativeModuleEnv::GetCacheUsage(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  Local<Context> context = env->context();
+  Local<Object> result = Object::New(isolate);
+  result
+      ->Set(env->context(),
+            OneByteString(isolate, "compiledWithCache"),
+            ToJsSet(context, env->native_modules_with_cache))
+      .FromJust();
+  result
+      ->Set(env->context(),
+            OneByteString(isolate, "compiledWithoutCache"),
+            ToJsSet(context, env->native_modules_without_cache))
+      .FromJust();
+  args.GetReturnValue().Set(result);
+}
+
+void NativeModuleEnv::ModuleIdsGetter(Local<Name> property,
+                                      const PropertyCallbackInfo<Value>& info) {
+  Isolate* isolate = info.GetIsolate();
+
+  std::vector<std::string> ids =
+      NativeModuleLoader::GetInstance()->GetModuleIds();
+  info.GetReturnValue().Set(
+      ToV8Value(isolate->GetCurrentContext(), ids).ToLocalChecked());
+}
+
+void NativeModuleEnv::ConfigStringGetter(
+    Local<Name> property, const PropertyCallbackInfo<Value>& info) {
+  info.GetReturnValue().Set(GetConfigString(info.GetIsolate()));
+}
+
+void NativeModuleEnv::RecordResult(const char* id,
+                                   NativeModuleLoader::Result result,
+                                   Environment* env) {
+  if (result == NativeModuleLoader::Result::kWithCache) {
+    env->native_modules_with_cache.insert(id);
+  } else {
+    env->native_modules_without_cache.insert(id);
+  }
+}
+void NativeModuleEnv::CompileFunction(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args[0]->IsString());
+  node::Utf8Value id_v(env->isolate(), args[0].As<String>());
+  const char* id = *id_v;
+  NativeModuleLoader::Result result;
+  MaybeLocal<Function> maybe =
+      NativeModuleLoader::GetInstance()->CompileAsModule(
+          env->context(), id, &result);
+  RecordResult(id, result, env);
+  if (!maybe.IsEmpty()) {
+    args.GetReturnValue().Set(maybe.ToLocalChecked());
+  }
+}
+
+// Returns Local<Function> of the compiled module if return_code_cache
+// is false (we are only compiling the function).
+// Otherwise return a Local<Object> containing the cache.
+MaybeLocal<Function> NativeModuleEnv::LookupAndCompile(
+    Local<Context> context,
+    const char* id,
+    std::vector<Local<String>>* parameters,
+    Environment* optional_env) {
+  NativeModuleLoader::Result result;
+  MaybeLocal<Function> maybe =
+      NativeModuleLoader::GetInstance()->LookupAndCompile(
+          context, id, parameters, &result);
+  if (optional_env != nullptr) {
+    RecordResult(id, result, optional_env);
+  }
+  return maybe;
+}
+
+void NativeModuleEnv::Initialize(Local<Object> target,
```

Comment:
Maybe
```suggestion
void NativeModuleEnv::InitializeTarget(Local<Object> target,
```
Since it initializes the target, not itself.

Reply:
I believe `Class::Initialize` is our convention everywhere in the code base? Searching for `::Initialize(` gives me 42 results in 20 files.

Reply:
The difference I see is `NativeModuleEnv` is a class and `Initialize` is a method, so reading this I assumed this applies to `NativeModuleEnv`, and that `target` is just an in-parameter.

But comment is non-blocking.

Reply:
In cases like this I would usually go with whatever the convention currently is, and change them all later in another patch if the other option makes more sense - I think consistency is more important here (but yes we should unify how these functions are named, I agree `Initialize` alone is a bit confusing, I was confused when I first saw it in the crypto classes.)